document/core/exec/instructions.rst

```diff
@@ -801,6 +801,93 @@ Memory Instructions
      (\otherwise)
    \end{array}
 
+.. _exec-memory.copy:
+
+:math:`\MEMORYCOPY`
+...................
+
+1. Assert: due to :ref:`validation <valid-memory.copy>`, a value of :ref:`value type <syntax-valtype>` |I32| is on the top of the stack.
+
+2. Pop the value :math:`\I32.\CONST~cnt` from the stack.
+
+3. Assert: due to :ref:`validation <valid-memory.copy>`, a value of :ref:`value type <syntax-valtype>` |I32| is on the top of the stack.
+
+4. Pop the value :math:`\I32.\CONST~src` from the stack.
+
+5. Assert: due to :ref:`validation <valid-memory.copy>`, a value of :ref:`value type <syntax-valtype>` |I32| is on the top of the stack.
+
+6. Pop the value :math:`\I32.\CONST~dst` from the stack.
+
+7. If :math:`cnt = 0`, then:
+
+    a. Return.
+
+8. If :math:`dst < src`, then:
+
+    a. Push the value :math:`\I32.\CONST~dst` to the stack.
+
+    b. Push the value :math:`\I32.\CONST~src` to the stack.
+
+    c. Execute the instruction :math:`\I32\K{.}\LOAD\K{8\_u}~\{ \OFFSET~0, \ALIGN~0 \}`.
+
+    d. Execute the instruction :math:`\I32\K{.}\STORE\K{8}~\{ \OFFSET~0, \ALIGN~0 \}`.
+
+    e. Push the value :math:`\I32.\CONST~(dst+1)` to the stack.
+
+    f. Push the value :math:`\I32.\CONST~(src+1)` to the stack.
+
+9. Else:
+
+    a. Push the value :math:`\I32.\CONST~(dst+cnt-1)` to the stack.
+
+    b. Push the value :math:`\I32.\CONST~(src+cnt-1)` to the stack.
+
+    c. Execute the instruction :math:`\I32\K{.}\LOAD\K{8\_u}~\{ \OFFSET~0, \ALIGN~0 \}`.
+
+    d. Execute the instruction :math:`\I32\K{.}\STORE\K{8}~\{ \OFFSET~0, \ALIGN~0 \}`.
+
+    e. Push the value :math:`\I32.\CONST~dst` to the stack.
+
+    f. Push the value :math:`\I32.\CONST~src` to the stack.
+
+10. Push the value :math:`\I32.\CONST~(cnt-1)` to the stack.
+
+11. Execute the instruction :math:`\MEMORYCOPY`.
+
+.. math::
+   ~\\[-1ex]
+   \begin{array}{l}
+   \begin{array}{lcl@{\qquad}l}
+   S; F; (\I32.\CONST~dst)~(\I32.\CONST~src)~(\I32.\CONST~0)~\MEMORYCOPY &\stepto& S; F; \epsilon
+   \end{array}
+   \\[1ex]
+   \begin{array}{lcl@{\qquad}l}
+   S; F; (\I32.\CONST~dst)~(\I32.\CONST~src)~(\I32.\CONST~(cnt+1))~\MEMORYCOPY &\stepto& S; F;
+     \begin{array}[t]{@{}l@{}}
+     (\I32.\CONST~dst) \\
+     (\I32.\CONST~src)~(\I32\K{.}\LOAD\K{8\_u}~\{ \OFFSET~0, \ALIGN~0 \}) \\
+     (\I32\K{.}\STORE\K{8}~\{ \OFFSET~0, \ALIGN~0 \}) \\
+     (\I32.\CONST~(dst+1))~(\I32.\CONST~(src+1))~(\I32.\CONST~cnt)~\MEMORYCOPY \\
```

Comment:
Crap, one annoyance I just realised is that dst+1 (and src+1) may overflow the i32 range, in which case `(i32.const (dst+1))` is not well-formed syntactically. That problem also applies to all other bulk operators. I think that requires a modification, e.g.:
```
S; F; (i32.const dst) (i32.const src) (i32.const cnt) memory.copy ~>
S; F; (i32.const dst) (i32.const src) (i32.load8_u {..}) (i32.store8 {...})
      const_i32(dst+1) const_i32(dst+1) (i32.const cnt) memory.copy
```
with the auxiliary meta function
```
const_t(x) = (t.const x)  (iff x is well-formed for t)
const_t(x) = trap  (otherwise)
```
Not sure where to best put that function, since it will be needed for the other bulk ops as well. Perhaps it could go into a "Conventions" subsection for the Values section in runtime.rst.

Reply:
Actually, the trap trick would be wrong when cnt happens to be 0... Hm, I can't think of a way to avoid an extra rule.

Reply:
FWIW, SpiderMonkey goes out of its way to avoid trapping just because a computation overflows, dropping into 64-bit arithmetic in some cases.  This goes for all the bulk ops.  Mostly this is to be able to trap midway through an operation though, without doing bounds checking (and checking for overflow) for every element.

Reply:
Ah, good point. I'll add it.

Reply:
@rossberg not sure what you mean, what's wrong with `cnt` = 0?

Reply:
The problem for cnt=0 is that this formulation now unconditionally traps after processing address 2^32-1, even when the remaining cnt is zero after that, which should be okay. So e.g. `(memory.copy (i32.const 2^32-1) (i32.const 0) (i32.const 1))` would incorrectly trap after writing the byte.

Unfortunately, I can't think of a way of handling that correctly without multiple extra rules, essentially duplicating the logic of the existing cases. The only other alternative is introducing an administrative version of each bulk instruction that incorporates the running indices as plain natural numbers and thus side-steps the value range limitation. Both solutions seem really annoying.

It's also unfortunate that it's practically impossible to add tests for these edge cases, even though engines will likely run into similar issues, as @lars-t-hansen pointed out.

Reply:
I see what you mean. I've added a `cnt=1` case for this, which I think works OK without being too verbose. PTAL